lib/pbench/server/api/resources/users_api.py

```diff
@@ -0,0 +1,534 @@
+            self.logger.warning("Missing username field")
+            abort(
+                400, message="Missing username field",
+            )
+        username = username.lower()
+        if username == "admin":
+            self.logger.warning("User tried to register with admin username")
+            abort(
+                400, message="Please choose another username",
+            )
+
+        # check if provided username already exist
+        try:
+            user = User.query(username=user_data.get("username"))
+        except Exception:
+            self.logger.exception("Exception while querying user")
+            abort(500, message="INTERNAL ERROR")
+        if user:
+            self.logger.warning(
+                "A user tried to re-register. Username: {}", user.username
+            )
+            abort(403, message="Provided username is already in use.")
+
+        password = user_data.get("password")
+        if not password:
+            self.logger.warning("Missing password field")
+            abort(
+                400, message="Missing password field",
+            )
+
+        email_id = user_data.get("email")
+        if not email_id:
+            self.logger.warning("Missing email field")
+            abort(
+                400, message="Missing email field",
+            )
+        # check if provided email already exist
+        try:
+            user = User.query(email=user_data.get("email_id"))
+        except Exception:
+            self.logger.exception("Exception while querying user")
+            abort(500, message="INTERNAL ERROR")
+        if user:
+            self.logger.warning("A user tried to re-register. Email: {}", user.email)
+            abort(403, message="Provided email is already in use")
+
+        first_name = user_data.get("first_name")
+        if not first_name:
+            self.logger.warning("Missing firstName field")
+            abort(
+                400, message="Missing firstName field",
+            )
+
+        last_name = user_data.get("last_name")
+        if not last_name:
+            self.logger.warning("Missing lastName field")
+            abort(
+                400, message="Missing lastName field",
+            )
+
+        try:
+            user = User(
+                username=username,
+                password=password,
+                first_name=first_name,
+                last_name=last_name,
+                email=email_id,
+            )
+
+            # insert the user
+            user.add()
+            self.logger.info(
+                "New user registered, username: {}, email: {}", username, email_id
+            )
+
+            response_object = {
+                "status": "success",
+                "message": "Successfully registered.",
+            }
+            response = jsonify(response_object)
+            response.status_code = 201
+            return make_response(response, 201)
+        except EmailNotValidError:
```

Comment:
It will be invoked when we try to add a user entry with an email field in it. We have defined a sqlalchemy `@validates("email")` decorator to validate email field in our user model and it validates the email whenever a new user entry is being processed.